### PR TITLE
Fix the short accounting

### PR DESCRIPTION
**Pull request by jalextowle**
The primary objective of this PR was to fix the short accounting. I believe that this has been accomplished (for an initial version of the short accounting); however, the logic still needs to be well tested. This will be one of the focuses next week, so I am leaving that out of this PR.

An ancillary change made during this PR was to add pyright as part of our linting CI. As of right now, this has been marked as `continue-on-error: true`, but it should ultimately be enforced.

### CODE REVIEW

File: examples/example_main.py [hunk 1]
```diff
@@ -36,8 +36,8 @@ def __init__(self, wallet_address: int, budget: int = 10_000) -> None:
 
     def action(self, market: Market) -> list[Any]:
         """Implement a custom user strategy"""
-        block_position_list = list(self.wallet.shorts.values())
-        has_opened_short = bool(any((x < -1 for x in block_position_list)))
+        shorts = list(self.wallet.shorts.values())
+        has_opened_short = bool(any((short.balance > 0 for short in shorts)))
```
Comment (dpaiton) [inline examples/example_main.py:40]: Cool I was just looking at this and I have no idea what that logic was doing before lol. I fixed it differently in my current workflow, but this is good I think.

File: src/elfpy/wallet.py [hunk 2]
```diff
@@ -25,12 +60,10 @@ class Wallet:
         The base assets that held by the trader.
     lp_tokens : float
         The LP tokens held by the trader.
-    longs : dict
+    longs : Dict[float, Long]
```
Comment (sentilesdal) [inline src/elfpy/wallet.py:63]: why is this `Dict[float, Long]`?  Shouldn't it be num or str for the key?

Comment (sentilesdal) [reply src/elfpy/wallet.py:63]: ah nvm, i see that the mint time is a float

File: src/elfpy/agent.py [hunk 3]
```diff
@@ -106,36 +106,68 @@ def update_wallet(self, wallet_deltas: Wallet, market: Market) -> None:
...
-                        if mint_time in self.wallet[key]:  #  entry already exists for this mint_time, so add to it
-                            self.wallet[key][mint_time] += amount
+                        if mint_time in self.wallet.longs:  #  entry already exists for this mint_time, so add to it
+                            self.wallet.longs[mint_time].balance += long.balance
                         else:
-                            self.wallet[key].update({mint_time: amount})
-                    if self.wallet[key][mint_time] == 0:
-                        del self.wallet[key][mint_time]
+                            self.wallet.longs.update({mint_time: long})
+                    if self.wallet.longs[mint_time].balance == 0:
```
Comment (sentilesdal) [inline src/elfpy/agent.py]: as a protection against rounding errors, maybe have this be `balance <= 0`

File: src/elfpy/wallet.py [hunk 4]
```diff
@@ -12,6 +12,41 @@
...
 
 
+@dataclass
+class Long:
+    """An open long position.
+
+    Arguments
+    ---------
+    balance : float
+        The amount of longs that are open.
```
Comment (sentilesdal) [inline src/elfpy/wallet.py:22]: this is a weird statement, is the balance in base or in # of bonds?

File: src/elfpy/wallet.py [hunk 5]
```diff
@@ -12,6 +12,41 @@
...
+
+
+@dataclass
+class Short:
+    """An open short position.
+
+    Arguments
+    ---------
+    balance : float
+        The amount of shorts that are open.
```
Comment (sentilesdal) [inline src/elfpy/wallet.py:38]: same here, weird sentence.  I assume these are in bonds.  margin below is in base?

**Review by sentilesdal — APPROVED** [review]
accounting looks right to me

**Pull request merged**